README.md
# Puppet-gitlab [![Build Status](https://travis-ci.org/sbadia/puppet-gitlab.png?branch=master)](https://travis-ci.org/sbadia/puppet-gitlab)

Tested successfully with Gitlab 6-8-stable on Ubuntu 12.04 and Debian Wheezy (7.2) with Puppet 3.2 or newer.

#### Table of contents

1. [Overview](#overview)
2. [Module description](#module-description)
3. [Parameters](#parameters)
4. [Usage](#usage)
    * [Basic usage](#basic-usage)
    * [With LDAP](#with-ldap)
5. [Limitation](#limitation)
6. [Development](#development)

# Overview

[GitLab](http://gitlab.org/) is a free project and repository management application

A [Puppet Module](http://docs.puppetlabs.com/learning/modules1.html#modules) is a collection of related content that can be used to model the configuration of a discrete service.

# Module description

This module is based on the admin guides for [gitlab](https://github.com/gitlabhq/gitlabhq/wiki), stable version.

- [puppet-gitlab](http://forge.puppetlabs.com/sbadia/gitlab) on puppet forge.

## Dependencies
- [puppetlabs/puppetlabs-stdlib](https://github.com/puppetlabs/puppetlabs-stdlib)
- [puppetlabs/puppetlabs-vcsrepo](https://github.com/puppetlabs/puppetlabs-vcsrepo)
- [puppetlabs/puppetlabs-git](https://github.com/puppetlabs/puppetlabs-git)

See [gitlab example](https://github.com/sbadia/vagrant-gitlab/blob/master/examples/gitlab.pp).

## GitLab web interface
- access via your browser under the hostname (e.g. http://gitlab.domain.tld)
- **Login**: admin@local.host
- **Password**: 5iveL!fe

1. Add an ssh key to your account, or create another account
2. Create a project
3. Play !

# Parameters

* `ensure`:  Ensure gitlab/gitlab-shell repo are present, latest. absent is not yet supported (default: present)
* `git_user`: Name of the gitlab (default: git)
* `git_group`: Name of the group for the gitlab user (default: $git_user)
* `git_home`: Home directory for gitlab repository (default: /home/git)
* `git_email`: Email address for gitlab user (default: git@someserver.net)
* `git_comment`: Gitlab user comment (default: GitLab)
* `gitlab_manage_user`: Whether to manage the user account for gitlab (default: true)
* `gitlab_manage_home`: Whether to manage the home directory for gitlab (default: true)
* `gitlab_sources`: Gitlab sources (default: git://github.com/gitlabhq/gitlabhq.git)
* `gitlab_branch`: Gitlab branch (default: 6-9-stable)
* `gitlabshell_sources`: Gitlab-shell sources (default: git://github.com/gitlabhq/gitlab-shell.git)
* `gitlabshell_branch`: Gitlab-shell branch (default: v1.9.4)
* `gitlab_http_port`: Port that NGINX listens on for HTTP traffic (default: 80)
* `gitlab_ssl_port`: Port that NGINX listens on for HTTPS traffic (default: 443)
* `gitlab_http_timeout`: HTTP timeout in seconds (unicorn/nginx) (default: 60)
* `gitlab_redishost`: Redis host used for Sidekiq (default: localhost)
* `gitlab_redisport`: Redis host used for Sidekiq (default: 6379)
* `gitlab_dbtype`: Gitlab database type (default: mysql)
* `gitlab_dbname`: Gitlab database name (default: gitlab\_db)
* `gitlab_dbuser`: Gitlab database user (default: gitlab\_user)
* `gitlab_dbpwd`: Gitlab database password (default: changeme)
* `gitlab_dbhost`: Gitlab database host (default: localhost)
* `gitlab_dbport`: Gitlab database port (default: 3306)
* `gitlab_domain`: Gitlab domain (default $fqdn)
* `gitlab_repodir`: Gitlab repository directory (default: $git\_home)
* `gitlab_backup`: Whether to enable automatic backups (default: false)
* `gitlab_backup_path`: Path where Gitlab's backup rake task puts its files (default: tmp/backups)
* `gitlab_backup_keep_time`: Retention time of Gitlab's backups (in seconds) (default: 0 == forever)
* `gitlab_backup_time`: Time (hour) when the Gitlab backup task is run from cron (default: fqdn\_rand(5)+1)
* `gitlab_backup_postscript`: Path to one or more shell scripts to be executed after the backup (default: false)
* `gitlab_relative_url_root`: Run GitLab in a non-root path (default: false, dont't forget the first slash)
* `gitlab_ssl`: Enable SSL for GitLab (default: false)
* `gitlab_ssl_cert`: SSL Certificate location (default: /etc/ssl/certs/ssl-cert-snakeoil.pem)
* `gitlab_ssl_key`: SSL Key location (default: /etc/ssl/private/ssl-cert-snakeoil.key)
* `gitlab_ssl_self_signed`: Set true if your SSL Cert is self signed (default: false)
* `gitlab_projects`: GitLab default number of projects for new users (default: 10)
* `gitlab_username_change`: Manage username changing in GitLab (default: true)
* `gitlab_unicorn_port`: Port that unicorn listens on 172.0.0.1 for HTTP traffic (default: 8080)
* `gitlab_unicorn_worker`: Number of unicorn workers (default: 2)
* `gitlab_bundler_flags`: Flags to be passed to bundler when installing gems (default: --deployment)
* `gitlab_bundler_jobs`: The number of jobs to use while installing gems. Should match number of CPUs on machine (default: 1)
* `gitlab_ensure_postfix`: Whether or not this module should ensure the postfix
  package is installed (used to manage conflicts with other modules) (default:
true)
* `exec_path`: PATH of executtion (default: `/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin`)
* `ldap_enabled`: Enable LDAP backend for gitlab web (see bellow) (default: false)
* `ldap_host`: FQDN of LDAP server (default: ldap.domain.com)
* `ldap_base`: LDAP base dn (default: dc=domain,dc=com)
* `ldap_uid`: Uid for LDAP auth (default: uid)
* `ldap_user_filter`: RFC 4515 style filter for the user (default: '')
* `ldap_port`: LDAP port (default: 636)
* `ldap_method`: Method to use (default: ssl)
* `ldap_bind_dn`: User for LDAP bind auth (default: nil)
* `ldap_bind_password`: Password for LDN bind auth (default: nil)
* `git_package_name`: Package name for git (default: git-core)
* `git_proxy`: Proxy for GIT access (default: undef)
* `ssh_port`: Port accepting SSH connections (default: 22)
* `google_analytics_id`: Google Analytics tracking ID (default: nil)
* `company_logo_url`: Url to the company logo to be displayed at the bottom of the sign_in page (default: '')
* `company_link`: Link to the company displayed under the logo of the company (default: '')
* `company_name`: Name of the company displayed under the logo of the company (default: '')
* `use_exim` : Apply a fix for compatibility with exim as explained at [gitlabhq/gitlabhq#4866](https://github.com/gitlabhq/gitlabhq/issues/486) (default: false)

# Usage

_Note:_ Assume that a database server is already installed on your server/infrastructure (see: [vagrant-gitlab](https://github.com/sbadia/vagrant-gitlab/blob/master/examples/gitlab.pp)).

## class gitlab

```puppet
class {
  'gitlab':
    git_email         => 'notifs@foobar.fr',
    git_comment       => 'GitLab',
    gitlab_domain     => 'gitlab.foobar.fr',
    gitlab_dbtype     => 'mysql',
    gitlab_dbname     => $gitlab_dbname,
    gitlab_dbuser     => $gitlab_dbuser,
    gitlab_dbpwd      => $gitlab_dbpwd,
    ldap_enabled      => false,
}
```

## class gitlab::ci

```puppet
class { 'gitlab::ci':
  ci_comment       => 'GitLab',
  gitlab_domain    => $gitlab_domain,
  gitlab_dbtype    => 'mysql',
  gitlab_dbname    => $ci_dbname,
  gitlab_dbuser    => $ci_dbuser,
  gitlab_dbpwd     => $ci_dbpwd,
  gitlab_http_port => 8081,
}
```

## class gitlab::ci::runner

```puppet
# The registration token can be found at: http://ci.example.com/admin/runners, accessible through Header > Runners.
class { 'gitlab::ci::runner':
  ci_server_url      => 'https://ci.example.com',
  registration_token => 'replaceme',
}

```

# Limitations

- TBF

# Development

Want to help - send a pull request.

## Development environment with vagrant

See [vagrant-gitlab](https://github.com/sbadia/vagrant-gitlab).


CHANGELOG
# vi: set ft=changelog :

2014-07-21 - 1.0.0
  * feature:
    - Major release
    - Manage ruby with rbenv (great thanks Andrew!!
      #114 (6 months, 48 comments))
    - Bump to gitlab 7.0 (6.9 → 7.0)
    - Allow users to disable curl package (thanks Dan)
    - Added parameters to manage jira/redmine forge (thanks Sam)

  * bugfix:
    - Fix bundler job flag with older bundler version (< 1.4)

  * others:
    - Added Andrew to contributors (full access on this repo)
    - Use semantic versionning http://semver.org/

2014-06-18 - 0.1.5
  * feature:
    - Use puppetlabs/git module for git package declaration
    - Add parameter for company link and logo support (thx Ludovic)
    - Fix compatibility issue with exim (thx Ludovic)
      https://github.com/gitlabhq/gitlabhq/issues/4866

  * bugfix:
    - Fix gitlab-satellites permissions (should be 0750)
    - Fix rspec output formatter (documentation)
    - Lock rspec version to 2.14.1 (puppet rspec not yet ready
      for RSpec +3.0, see https://github.com/rodjek/rspec-puppet/pull/204 )

2014-05-26 - 0.1.4
  * feature:
    - added ldap_user_filter parameter (RFC 4515 style filter for the
      user) (thanks Igor)
    - added nginx to listen ipv6 also
    - allow end users to disable nginx (with the param.
      manage_nginx) (thanks Andrew)
    - added support for nginx domain aliases (thanks Leonardo)
    - added gitlab_ensure_postfix parameter (to manage or not postfix package)
    - disable gzip compression if SSL enabled (nginx)
      - and enable it for static assets
    - bump to gitlab 6.9 + gitlab-shell 1.9.4 (6.7 → 6.8 → 6.9)
    - allow adjustment of number of bundler threads
    - simplify backup task

  * bugfix:
    - Fix travis gate (ruby1.8 and rake > 10.1.0)
    - remove MySQL reaping_frequency

2014-03-25 - 0.1.3
  * feature:
    - added ssh_port parameter (thanks Kalman)
    - added git_proxy parameter (thanks Stefan)
    - added google_analytics_id parameter (thanks Andrew)
    - internals unit-tests refactoring, better coverage and regexp
    - bump to GitLab 6.7 and GitLab Shell 1.9.1

  * bugfix:
    - allow special characters in db passwords (thanks Thomas)
    - fixed asset compilation and db migrations (thanks Thomas)

2014-02-22 - 0.1.2
  * feature:
    - manage gitlab relative URL (thanks Vincent)
    - add backups support + external script (thanks Igor)
    - bump to gitlab 6.6 (6.3 → 6.4 → 6.5 → 6.6)
    - securing SSL configuration (thanks Andrew, Igor)
    - allow « plain » for ldap_method (thanks sven)
    - manage http timeout and unicorn workers as parameter
    - manage exec_path as a parameter
    - replace git exec by vcsrepo module (thanks Igor)

  * bugfix:
    - internals: fix spec tests and travis config (thanks Lee)
    - fix git package name in RedHat* (thanks Stefan)
    - allow users to use non-stable GitLab branchs

2013-12-03 - 0.1.1
  * feature:
    - improve documentation (typos)

  * bugfix:
    - fix params in gitlab.yml (http/https with non-default port)
    - fix stdlib dependency (librarian require a version number)

2013-11-27 - 0.1.0
  * feature:
    - bump to GitLab 6.3 and gitlab-shell v1.7.9
    - add rack_attack and logrotate configurations

2013-11-17 - 0.0.10
  * feature:
    - bump gitlab-shell to 1.7.8 (multiple security fix)

  * bugfix:
    - bugfix https://github.com/sbadia/puppet-gitlab/pull/80

2013-11-08 - 0.0.9
  * feature:
    - huge changes/re-factorization by atomaka ! (many thanks !!)
    - Use anchors and refactoring of args/class
    - Add ssl support for nginx
    - Add extra params (repodir,username_changing,redis,unicorn)
    - Better management of extra packages (thx stdlib)
    - Bump to gitlab 6.2.3
    - Add spec and travis testing

2013-06-10 - 0.0.8
  * feature:
    - Use nginx,ruby,redis,mysql external modules
    - Clean pre.pp file

2013-04-27 - 0.0.6
  * feature:
    - Remove apt and mysql setting from core module
    - Bump to GitLab 5.1 (switch from unicorn to puma)

  * bugfix:
    - Fix packaging issue (https://github.com/sbadia/puppet-gitlab/issues/33) wait for
      a cleaner way to do that (http://projects.puppetlabs.com/issues/14651)
    - Fix timeouts issue, and others bugs

2013-04-07 - 0.0.5
  * feature:
    - up to GitLab 5.0
    - remove gitolite (use gitlab-shell)

2012-11-02 - 0.0.4
  * feature:
    - up to GitLab 4.1

2013-01-01 - 0.0.3
  * feature:
    - up to GitLab 3.2
    - add dependency to mysql, stdlib

2012-08-12 - 0.0.1
  * feature:
    - initial release


spec/spec_helper.rb
require 'puppetlabs_spec_helper/module_spec_helper'
require 'shared_examples'
#require 'simplecov'
#
#SimpleCov.start do
#    add_filter "/spec/"
#end

RSpec.configure do |c|
  c.alias_it_should_behave_like_to :it_configures, 'configures'
  c.alias_it_should_behave_like_to :it_raises, 'raises'

  c.default_facts = {
    :osfamily               => 'Debian',
    :operatingsystem        => 'Debian',
    :kernel                 => 'Linux',
    :lsbdistid              => 'debian',
    :lsbdistcodename        => 'wheezy',
    :operatingsystemrelease => '6.5',
    :fqdn                   => 'gitlab.fooboozoo.fr',
    :sshrsakey              => 'AAAAB3NzaC1yc2EAAAA',
    :concat_basedir         => '/var/lib/puppet/concat'
  }

end


spec/gitlab_shared.rb
shared_context "gitlab_shared", :a => :b do

  def global_gitlab_variables
   "$git_user           = 'git'
    $git_home           = '/home/git'
    $git_email          = 'git@someserver.net123'
    $git_comment        = 'git version control'
    $git_admin_pubkey   = '#Not configured'
    $git_admin_privkey  = '#Not configured'
    $ssh_key_provider   = 'source'
    $gitlab_user        = 'gitlab'
    $gitlab_home        = '/home/gitlab'
    $gitlab_comment     = 'gitlab system'
    $gitlab_sources     = 'git://github.com/gitlabhq/gitlabhq.git'
    $gitlab_branch      = 'stable'
    $gitolite_sources   = 'git://github.com/gitlabhq/gitolite.git'
    $gitolite_branch    = 'gl-v320'
    $gitlab_dbtype      = 'mysql'
    $gitlab_dbname      = 'gitladb'
    $gitlab_dbuser      = 'gitladbu'
    $gitlab_dbpwd       = 'change-me123'
    $ldap_enabled       = false
    $ldap_host          = 'ldap.domain.com'
    $ldap_base          = 'dc=domain,dc=com'
    $ldap_uid           = 'uid'
    $ldap_port          = '636'
    $ldap_method        = 'ssl'
    $ldap_bind_dn       = ''
    $ldap_bind_password = ''"
  end

end


spec/shared_examples.rb
shared_examples_for "a Puppet::Error" do |description|
  it "with message matching #{description.inspect}" do
    expect { should have_class_count(1) }.to raise_error(Puppet::Error, description)
  end
end


files/etc/init.d/sidekiq.sh
#! /bin/bash

# Sidekiq
# Maintainer: @randx
# App Version: 5.2

### BEGIN CHKCONFIG INFO
# chkconfig 2345 95 30
#### END CHKCONFIG INFO

### BEGIN INIT INFO
# Provides:          gitlab
# Required-Start:    $local_fs $remote_fs $network $syslog redis-server
# Required-Stop:     $local_fs $remote_fs $network $syslog
# Default-Start:     2 3 4 5
# Default-Stop:      0 1 6
# Short-Description: Sidekiq is a backend helper for GitLab
# Description:       Sidekiq is a backend helper for GitLab
### END INIT INFO


APP_ROOT="/home/git/gitlab"
APP_USER="git"
PID_PATH="$APP_ROOT/tmp/pids"
SOCKET_PATH="$APP_ROOT/tmp/sockets"
SIDEKIQ_PID="$PID_PATH/sidekiq.pid"
STOP_SIDEKIQ="RAILS_ENV=production bundle exec rake sidekiq:stop"
START_SIDEKIQ="RAILS_ENV=production bundle exec rake sidekiq:start"
NAME="sidekiq"
DESC="Sidekiq service"

check_pid(){
  if [ -f $SIDEKIQ_PID ]; then
    PID=`cat $SIDEKIQ_PID`
    STATUS=`ps aux | grep $PID | grep -v grep | wc -l`
  else
    STATUS=0
    PID=0
  fi
}

execute() {
  sudo -u $APP_USER -H bash -l -c "$1"
}

start() {
  cd $APP_ROOT
  check_pid
  if [ "$PID" -ne 0 -a "$STATUS" -ne 0 ]; then
    # Program is running, exit with error code 1.
    echo "Error! $DESC $NAME is currently running!"
    exit 1
  else
    if [ `whoami` = root ]; then
      execute "mkdir -p $PID_PATH && $START_SIDEKIQ  > /dev/null  2>&1 &"
      echo "$DESC started"
    fi
  fi
}

stop() {
  cd $APP_ROOT
  check_pid
  if [ "$PID" -ne 0 -a "$STATUS" -ne 0 ]; then
    ## Program is running, stop it.
    execute "mkdir -p $PID_PATH && $STOP_SIDEKIQ  > /dev/null  2>&1 &"
    echo "$DESC stopped"
  else
    ## Program is not running, exit with error.
    echo "Error! $DESC not started!"
    exit 1
  fi
}

restart() {
  cd $APP_ROOT
  check_pid
  if [ "$PID" -ne 0 -a "$STATUS" -ne 0 ]; then
    stop
    start
    echo "$DESC restarted."
  else
    echo "$NAME not running!"
    start
  fi
}

status() {
  cd $APP_ROOT
  check_pid
  if [ "$PID" -ne 0 -a "$STATUS" -ne 0 ]; then
    echo "$DESC with PID $PID is running."
  else
    echo "$DESC is not running."
    exit 1
  fi
}

## Check to see if we are running as root first.
## Found at http://www.cyberciti.biz/tips/shell-root-user-check-script.html
if [ "$(id -u)" != "0" ]; then
    echo "This script must be run as root"
    exit 1
fi

case "$1" in
  start)
        start
        ;;
  stop)
        stop
        ;;
  restart)
        restart
        ;;
  reload|force-reload)
        echo -n "Reloading $NAME configuration: "
        kill -HUP `cat $PID`
        echo "done."
        ;;
  status)
        status
        ;;
  *)
        echo "Usage: sudo service sidekiq {start|stop|restart|reload}" >&2
        exit 1
        ;;
esac

exit 0

files/etc/init.d/gitlab.sh
#! /bin/bash
# GITLAB
# Maintainer: @randx
# App Version: 5.2

### BEGIN CHKCONFIG INFO
# chkconfig 2345 70 29
#### END CHKCONFIG INFO

### BEGIN INIT INFO
# Provides:          gitlab
# Required-Start:    $local_fs $remote_fs $network $syslog redis-server
# Required-Stop:     $local_fs $remote_fs $network $syslog
# Default-Start:     2 3 4 5
# Default-Stop:      0 1 6
# Short-Description: GitLab git repository management
# Description:       GitLab git repository management
### END INIT INFO


APP_ROOT="/home/git/gitlab"
APP_USER="git"
DAEMON_OPTS="-C $APP_ROOT/config/puma.rb"
PID_PATH="$APP_ROOT/tmp/pids"
SOCKET_PATH="$APP_ROOT/tmp/sockets"
WEB_SERVER_PID="$PID_PATH/puma.pid"
NAME="gitlab"
DESC="GitLab service"

check_pid(){
  if [ -f $WEB_SERVER_PID ]; then
    PID=`cat $WEB_SERVER_PID`
    STATUS=`ps aux | grep $PID | grep -v grep | wc -l`
  else
    STATUS=0
    PID=0
  fi
}

execute() {
  sudo -u $APP_USER -H bash -l -c "$1"
}

start() {
  cd $APP_ROOT
  check_pid
  if [ "$PID" -ne 0 -a "$STATUS" -ne 0 ]; then
    # Program is running, exit with error code 1.
    echo "Error! $DESC $NAME is currently running!"
    exit 1
  else
    if [ `whoami` = root ]; then
      if [ -S "$SOCKET_PATH/gitlab.socket" ]; then
        execute "rm $SOCKET_PATH/gitlab.socket"
      fi
      execute "RAILS_ENV=production bundle exec puma $DAEMON_OPTS"
      echo "$DESC started"
    fi
  fi
}

stop() {
  cd $APP_ROOT
  check_pid
  if [ "$PID" -ne 0 -a "$STATUS" -ne 0 ]; then
    ## Program is running, stop it.
    kill -QUIT `cat $WEB_SERVER_PID`
    rm "$WEB_SERVER_PID" >> /dev/null
    echo "$DESC stopped"
  else
    ## Program is not running, exit with error.
    echo "Error! $DESC not started!"
    if [ -S "$SOCKET_PATH/gitlab.socket" ]; then
      execute "rm $SOCKET_PATH/gitlab.socket"
    fi     
    exit 1
  fi
}

restart() {
  cd $APP_ROOT
  check_pid
  if [ "$PID" -ne 0 -a "$STATUS" -ne 0 ]; then
    echo "Restarting $DESC..."
    kill -USR2 `cat $WEB_SERVER_PID`
    echo "$DESC restarted."
  else
    echo "Error, $NAME not running!"
    exit 1
  fi
}

status() {
  cd $APP_ROOT
  check_pid
  if [ "$PID" -ne 0 -a "$STATUS" -ne 0 ]; then
    echo "$DESC / Puma with PID $PID is running."
  elif [ -S "$SOCKET_PATH/gitlab.socket" ]; then
    echo "${DESC} is not running, but ${SOCKET_PATH}/gitlab.socket exists"
    exit 1
  else
    echo "$DESC is not running."
    exit 1
  fi
}

## Check to see if we are running as root first.
## Found at http://www.cyberciti.biz/tips/shell-root-user-check-script.html
if [ "$(id -u)" != "0" ]; then
    echo "This script must be run as root"
    exit 1
fi

case "$1" in
  start)
        start
        ;;
  stop)
        stop
        ;;
  restart)
        restart
        ;;
  reload|force-reload)
        echo -n "Reloading $NAME configuration: "
        kill -HUP `cat $PID`
        echo "done."
        ;;
  status)
        status
        ;;
  *)
        echo "Usage: sudo service gitlab {start|stop|restart|reload}" >&2
        exit 1
        ;;
esac

exit 0